Improve error handling for partial aggregation pushdown

rschlussel opened the pull request:
Improve error handling for partial aggregation pushdown and prevent returning wrong results when footer stats should not be relied on.  This covers the following cases:
1. Aggregations have been pushed down but partition file format does not support aggregation pushdown (can occur if table is declared with a supported storage format, but partition has a different storage format). Previously, page source providers for some file formats had special handling for this case, but not all
2. Always throw an exception if aggregations have been pushed down but partition footer stats are unreliable. Previously, if filter pushdown was enabled (used OrcSelectivePageSourceFactory), we wouldn't create an AggregatedPageSource, so you would get an error somewhere on read. If it was disabled (OrcBatchPageSourceFactory), we would create an AggregatedPageSource and the query would silently give wrong results.
3. Unexpected state where some but not all columns are of AGGREGATED type.

Error handling is still going to be reader dependent if both the table and partition format support partial aggregation pushdown, but the partition format does not support as many types (e.g. currently supports more types for partial aggregation pushdown).

## Description
Previously AggregatedPageSources (which support the execution side of partial aggregation pushdown) were created from within the selective and batch page source factories of supported file formats. Similarly error handling for any unsupported file format needed to be repeated for each PageSourceFactory of all unsupported file formats.  This resulted in a fragmented implementation and some unsupported file formats that did not include proper error handling.

Additionally, partial aggregation pushdown cannot be used when footer stats are unreliable, however handling for this was only added for one of the supported file formats factories (OrcSelectivePageSourceFactory) while others (orc and parquet batch factories) could silently return wrong results.  Furthermore, the handling in OrcSelectivePageSourceFactory prevented wrong results by not creating an aggregated page source but didn't produce a clear error message because it kept going by trying to create a selective page source.

This PR makes HiveAggregatedPageSourceFactories into a top-level concept similar to HiveSelectivePageSourceFactories and HiveBatchPageSourceFactories so that we can unify all the error handling and prevent bugs from creeping in as new file format page source factories are added.  
The main logic of the change is in HivePageSourceProvider.  A lot of the rest of it is scaffolding to support that.

## Motivation and Context:
1. to ensure consistent error handling across different page sources even as new page formats or selective readers implementations are added.
2. To prevent wrong results when footer stats are unreliable regardless of file format and any other configs.

This gap was discovered as part of an audit to make sure we were not assuming that partition file formats will always match table file formats.

## Impact
Fix a potential wrong results bug  when footer stats are marked as unreliable and aggregation pushdown is enabled.  Ensure all file formats that don't support aggregation pushdown will return a clear error to the user. 

## Test Plan
new unit tests for HivePageSourceProvider

## Contributor checklist

- [X] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [X] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [X] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [X] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [X] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==
Hive Changes
* Fix a potential wrong results bug when footer stats are marked unreliable and partial aggregation pushdown is enabled.  Such queries will now fail with an error.
```

rschlussel commented:
I'm still updating the tests. Don't review yet

rschlussel commented:
this is ready for review (failing tests are flaky/unrelated)

abhiseksaikia reviewed (dismissed):
LGTM % minor nit and a question

rschlussel commented on presto-hive/src/main/java/com/facebook/presto/hive/orc/DwrfAggregatedPageSourceFactory.java:
```diff
@@ -0,0 +1,133 @@
+            Optional<EncryptionInformation> encryptionInformation,
+            boolean appendRowNumberEnabled)
+    {
+        if (!OrcSerde.class.getName().equals(storage.getStorageFormat().getSerDe())) {
+            return Optional.empty();
+        }
+
+        // per HIVE-13040 and ORC-162, empty files are allowed
+        if (fileSplit.getFileSize() == 0) {
+            return Optional.of(new FixedPageSource(ImmutableList.of()));
+        }
```
oh that's a bug! Dwrf files don't support empty files.

rschlussel replied:
fixed! thank you!

ClarenceThreepwood commented on presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java:
```diff
@@ -614,7 +653,8 @@ private static boolean shouldSkipBucket(HiveTableLayoutHandle hiveLayout, HiveSp
     }
 
-    private static boolean shouldSkipPartition(TypeManager typeManager, HiveTableLayoutHandle hiveLayout, DateTimeZone hiveStorageTimeZone, HiveSplit hiveSplit, SplitContext splitContext)
+    private static boolean shouldSkipPartition(TypeManager typeManager, HiveTableLayoutHandle hiveLayout, DateTimeZone hiveStorageTimeZone, HiveSplit hiveSplit, SplitContext
```
nit: this function signature and the one below
```suggestion
    private static boolean shouldSkipPartition(TypeManager typeManager, 
    HiveTableLayoutHandle hiveLayout, 
    DateTimeZone hiveStorageTimeZone, 
    HiveSplit hiveSplit, 
    SplitContext splitContext)
```

ClarenceThreepwood commented on presto-hive/src/main/java/com/facebook/presto/hive/orc/OrcAggregatedPageSourceFactory.java:
```diff
@@ -0,0 +1,259 @@
+        boolean lazyReadSmallRanges = getOrcLazyReadSmallRanges(session);
+
+        OrcDataSource orcDataSource;
+        Path path = new Path(fileSplit.getPath());
+        try {
+            FSDataInputStream inputStream = hdfsEnvironment.getFileSystem(session.getUser(), path, configuration).openFile(path, hiveFileContext);
+            orcDataSource = new HdfsOrcDataSource(
+                    new OrcDataSourceId(fileSplit.getPath()),
+                    fileSplit.getFileSize(),
+                    maxMergeDistance,
+                    maxBufferSize,
+                    streamBufferSize,
+                    lazyReadSmallRanges,
+                    inputStream,
+                    stats);
+        }
+        catch (Exception e) {
+            if (nullToEmpty(e.getMessage()).trim().equals("Filesystem closed") ||
+                    e instanceof FileNotFoundException) {
+                throw new PrestoException(HIVE_CANNOT_OPEN_SPLIT, e);
+            }
+            throw new PrestoException(HIVE_CANNOT_OPEN_SPLIT, splitError(e, path, fileSplit.getStart(), fileSplit.getLength()), e);
+        }
+
+        OrcAggregatedMemoryContext systemMemoryUsage = new HiveOrcAggregatedMemoryContext();
+        try {
+            DwrfKeyProvider dwrfKeyProvider = new ProjectionBasedDwrfKeyProvider(encryptionInformation, columns, useOrcColumnNames, path);
+            OrcReader reader = new OrcReader(
+                    orcDataSource,
+                    orcEncoding,
+                    orcFileTailSource,
+                    stripeMetadataSourceFactory,
+                    systemMemoryUsage,
+                    orcReaderOptions,
+                    hiveFileContext.isCacheable(),
+                    dwrfEncryptionProvider,
+                    dwrfKeyProvider,
+                    hiveFileContext.getStats());
+
+            List<HiveColumnHandle> physicalColumns = getPhysicalHiveColumnHandles(columns, useOrcColumnNames, reader.getTypes(), path);
```
+1

ClarenceThreepwood reviewed (approved):
Thanks for improving on the original implementation. Overall lgtm

rschlussel commented:
thanks for review @abhiseksaikia  and @ClarenceThreepwood. I've addressed your comments. I also split out the commits a bit as per request from @ajaygeorge.

ajaygeorge commented on presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java:
```diff
@@ -484,7 +454,7 @@ public static boolean checkSchemaMatch(org.apache.parquet.schema.Type parquetTyp
             case INT32:
                 return prestoType.equals(INTEGER) || prestoType.equals(BIGINT) || prestoType.equals(SMALLINT) || prestoType.equals(DATE) || prestoType.equals(DECIMAL) ||
-                    prestoType.equals(TINYINT) || prestoType.equals(REAL) || prestoType.equals(StandardTypes.DOUBLE);
+                        prestoType.equals(TINYINT) || prestoType.equals(REAL) || prestoType.equals(StandardTypes.DOUBLE);
```
stray space?

ajaygeorge reviewed (commented):
Consolidate error handling for ParquetPageSourceFactory https://github.com/prestodb/presto/pull/22011/commits/a8c2a388f8971cca2261be6e320e703604064712 looks good % a nit

ajaygeorge commented on presto-hive/src/main/java/com/facebook/presto/hive/rcfile/RcFilePageSourceFactory.java:
```diff
@@ -109,12 +108,6 @@ public Optional<? extends ConnectorPageSource> createPageSource(
             Optional<EncryptionInformation> encryptionInformation)
     {
-        if (!columns.isEmpty() && columns.stream().allMatch(hiveColumnHandle -> hiveColumnHandle.getColumnType() == AGGREGATED)) {
-            throw new UnsupportedOperationException("Partial aggregation pushdown only supported for ORC/Parquet files. " +
```
curious. where does this check move after the refactoring. I wasn't able to find it. Is not needed any more.?

rschlussel replied:
tagged you where this check is moved to.  instead of adding error handling for every file format, we do it all in one place. that's why it's not needed here anymore.

ajaygeorge commented on presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java:
```diff
@@ -373,8 +373,7 @@ private static Optional<ConnectorPageSource> createSelectivePageSource(
                     fileContext,
                     encryptionInformation,
-                    layout.isAppendRowNumberEnabled(),
-                    layout.isFooterStatsUnreliable());
```
Is footer stats reliable now?. The commit message doesn't clarify that.

ajaygeorge reviewed (commented):
Remove unneeded error handling from page source factories https://github.com/prestodb/presto/pull/22011/commits/f7fae208bb24f4aa2fce227c9ed54bd32f4791f8 looks good % some comments.

ajaygeorge reviewed (dismissed):
Rest commits look good. LGTM

ajaygeorge reviewed (approved):
LGTM

abhiseksaikia reviewed (approved):
LGTM!

ClarenceThreepwood reviewed (approved):
lgtm

sdruzkin commented on presto-hive/src/main/java/com/facebook/presto/hive/orc/OrcAggregatedPageSourceFactory.java:
```diff
@@ -0,0 +1,209 @@
+            DwrfEncryptionProvider dwrfEncryptionProvider,
+            boolean appendRowNumberEnabled)
+    {
+        OrcDataSource orcDataSource = getOrcDataSource(session, fileSplit, hdfsEnvironment, configuration, hiveFileContext, stats);
```
@rschlussel this is resource leak because we don't close the orcDataSource in a happy case

rschlussel replied:
oh good catch. Thank you!
